### PR TITLE
chore(precompile): account safety checks, generic interface refactor

### LINKED ISSUE #423: chore: Check for precompile module account on startup
Should verify that a module account exists for a given precompile at startup. Just as a safety measure against the 1 in a trillion chance some gets a priv key or deploys a contract.

### PR DESCRIPTION
**Pull request by calbera**
closes #423

### CODE REVIEW

File: cosmos/x/evm/plugins/precompile/plugin.go [hunk 1]
```diff
@@ -68,6 +70,21 @@ type plugin struct {
 
 // NewPlugin creates and returns a `plugin` with the default kv gas configs.
 func NewPlugin(precompiles []vm.RegistrablePrecompile) Plugin {
+	// verify every precompile contract's address is unique and corresponds to a specific Cosmos
+	// module account address, used by Polaris.
+	for _, precompile := range precompiles {
```
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/x/evm/plugins/precompile/plugin.go:75]: need a context to check the account keeper, don't have one during startup?

[the remark this answers is not in the document]
Comment (itsdevbear) [reply cosmos/x/evm/plugins/precompile/plugin.go:75]: do it in init genesis

**Comment by calbera** [comment]
@itsdevbear at time of InitGenesis, it seems the context is not setup? 
```
0xd9A998CaC66092748FfEc7cFBD155Aae1737C2fF
acc is nil


0x4381dC2aB14285160c808659aEe005D51255adD7
acc is nil


0xBDF49C3C3882102fc017FFb661108c63a836D065
acc is nil


0x93354845030274cD4bf1686Abd60AB28EC52e1a7
acc is nil


0x7b5Fe22B5446f7C62Ea27B8BD71CeF94e03f3dF2
acc is nil
```

**Comment by itsdevbear** [comment]
Check the genesis ordering, need to setup auth module before evm module

**Comment by calbera** [comment]
https://github.com/berachain/polaris/blob/da7eeae7aa0c0fe7de2ec6b69fea7f3d4dd45b1e/cosmos/runtime/config/runtime.go#L91-L98 auth is before evm

**Comment by itsdevbear** [comment]
Hmmm, try making sure it is propagating through the plugins correctly.

**Review by itsdevbear — COMMENTED** [review]
🥺

image

File: cosmos/x/evm/plugins/precompile/genesis.go [hunk 2]
```diff
@@ -22,10 +22,29 @@ package precompile
...
+	cosmlib "pkg.berachain.dev/polaris/cosmos/lib"
 	"pkg.berachain.dev/polaris/cosmos/x/evm/types"
 )
 
-func (p *plugin) InitGenesis(_ sdk.Context, _ *types.GenesisState) {}
+func (p *plugin) InitGenesis(ctx sdk.Context, _ *types.GenesisState) {
+	// ensure every precompile contract's address is unique and exists in the account keeper
+	for _, precompile := range p.precompiles {
+		if p.ak.GetAccount(ctx, cosmlib.AddressToAccAddress(precompile.RegistryKey())) == nil {
+			p.ak.SetAccount(
```
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/x/evm/plugins/precompile/genesis.go:35]: yeah makes more sense than genesis, its just it needs ctx to hit the keeper funcs

[the remark this answers is not in the document]
Comment (itsdevbear) [reply cosmos/x/evm/plugins/precompile/genesis.go:35]: Yeah just panic if they don't exist, but the creation should be triggered by them being registered in the app config and thus genesis.json

**Pull request closed without merging**